javalin/src/main/java/io/javalin/core/validation/Validator.kt

```diff
@@ -6,107 +6,44 @@
 
 package io.javalin.core.validation
 
+import io.javalin.core.util.JavalinLogger
 import io.javalin.http.BadRequestResponse
+import io.javalin.http.InternalServerErrorResponse
 
-open class Validator<T>(val value: T?, val messagePrefix: String = "Value", val key: String = "Parameter") {
+enum class ValidationError { NULLCHECK_FAILED, TYPE_CONVERSION_FAILED, CUSTOM_CHECK_FAILED, DESERIALIZATION_FAILED }
 
-    data class Rule<T>(val fieldName: String, val test: (T) -> Boolean, val invalidMessage: String)
-
-    protected open var errors: MutableMap<String, MutableList<String>>? = null
-
-    protected val rules = mutableSetOf<Rule<T>>()
-
-    @JvmOverloads
-    open fun check(predicate: (T) -> Boolean, errorMessage: String = "Failed check"): Validator<T> {
-        rules.add(Rule(key, predicate, errorMessage))
-        return this
-    }
-
-    //These two options will fail fast but only provide the first failure.
-    fun get(): T = getOrNull() ?: throw BadRequestResponse("$messagePrefix cannot be null or empty")
+/**
+ * The non-nullable [Validator] uses [Rule] rules, but checks if value is null before calling them.
+ * The [check] method wraps its non-nullable predicate in a nullable predicate
+ */
+open class Validator<T>(value: T?, val fieldName: String) : BaseValidator<T>(value) {
 
-    fun getOrNull(): T? {
-        if (value == null) return null
-        return rules.find { !it.test.invoke(value) }?.let { throw BadRequestResponse("$messagePrefix invalid - ${it.invalidMessage}") } ?: value
+    init {
+        addRule(fieldName, { it != null }, {
```

Comment:
Maybe instead of adding a rule override get? 
```kotlin
get(): T {
   return if (value == null) throw BadRequestResponse("$fieldName is required.") else super.get()
}
```

This would prevent your not null assertions that you add in check from throwing a null pointer. It would also allow you to pass any rules that were already created to the nullable validator without trying to filter this one out.